035-cluster-wide-volume-usage-quota-managemnet.md

```diff
@@ -0,0 +1,280 @@
+see [issue#2](https://github.com/strimzi/kafka-quotas-plugin/issues/2)) when considering whether to apply throttling to
+clients. This is problematic with respect
+to handling disk failure for JBODs [(KIP-112)](https://cwiki.apache.org/confluence/display/KAFKA/KIP-112%3A+Handle+disk+failure+for+JBOD) as the broker
+will take partitions offline when the volume they are stored on runs out of space. Which in the case of unbalanced usage
+between volumes can lead to a volume running out of storage without throttling being applied. The broker will go offline
+if all volumes are unavailable.
+
+Configuring the soft and hard limit thresholds is currently done in terms of bytes consumed, which forces users
+deploying the plugin to calculate a threshold for throttling and would require separate thresholds per volume to avoid
+out of space issues.
+
+## Proposal
+### Leverage Kafka to distribute volume usage metrics throughout the cluster.
+By publishing per volume usage metrics to a compacted topic keyed by the broker ID, each instance of the plugin will be able to start and quickly determine the state of the cluster and consistently apply throttling regardless of where the
+client is connected.
+
+[KIP-257](https://cwiki.apache.org/confluence/display/KAFKA/KIP-257+-+Configurable+Quota+Management) describes configurable quota management, where quotas are defined in terms of how much of a delay to apply to a given produce request. This leads to defining the
+possible states of the plugin as:
+`OPEN`, `THROTTLE` and `PAUSE` to accurately reflect the effects of transitioning between states.
+
+This implies there are three roles.
+1. **Data source** - Publish volume usage metrics about the volumes underlying the brokers log dirs.
+2. **Data Sink** - Consume volume usage metrics and make those available for making Quota Policy decisions.
+3. **Quota Policy** - Convert volume usage metrics for the entire cluster into throttling decisions.
+
+This proposal envisages them all being implemented by the kafka-static-quota-plugin however by clearly distinguishing the roles we leave open future options to move each of the roles out of process.
+
+#### Limit Types
+1. Consumed space limit: Triggers if the consumed space of a volume breaches the value. Candidate for deprecation.
+2. Minimum Free Bytes: Triggers if the amount of free space on a volume drops below the configured level. Particularly
+   useful for hard limits as an absolute minimum of free space.
+3. Minimum Free Percentage. Triggers if the amount of free space on the volume drops below the configured proportion of
+   the volumes total capacity.
+
+#### The data source should:
+- Collect local volume usage
+- Publish usage to an internal compacted topic keyed by broker ID.
+- Collect and publish volume usage metrics (using this [message schema](#Message-schema)) on start up and periodically after that.
```

Comment:
> Assuming there's nothing else on the volume, we can get the size of each replica via the `describeLogDirs()` Admin API. If we sum all sizes per logdir, we get the disk usage. So do we really need to emit this usage metric?

I'm not sure how valid that assumption is, probably fine for strimzi but not in general. While it would give us the usage per logDir we still wouldn't know how logDir's map to volumes and the capacities of each volume. So I think we still need to emit volume details. At which point I think its easier to just emit a single message with both consumed bytes and capacity in bytes.



> Also Cruise Control has a metrics reporter that write metrics in a topic to be used by other components. Have we considered reusing it?

The initial thought process was to do something which could stand on its own without rather than relying on external services. I.e. cruise control or a metrics service.